ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -0,0 +1,3018 @@
+    guard let keyboardDict = keyboardsDictionary[kbKey] else {
+      let error = NSError(domain: "Keyman", code: 0,
+                          userInfo: [NSLocalizedDescriptionKey: "Keyboard not found with key: \(kbKey)"])
+      downloadFailed(forKeyboard: kbInfo, error: error)
+      return
+    }
+
+    let urlStr = keyboardsInfo[keyboardID]![Key.keyboardURI]!
+    let keyboardURL = URL(string: urlStr)!
+
+    let kbFontStr = keyboardDict[Key.font] ?? ""
+    let kbOskFontStr = keyboardDict[Key.oskFont] ?? ""
+    var kbFontURLs: [URL] = []
+    do {
+      kbFontURLs += try keyboardFontURLs(jsonString: kbFontStr)
+      kbFontURLs += try keyboardFontURLs(jsonString: kbOskFontStr)
+    } catch {
+      downloadFailed(forKeyboard: kbInfo, error: error as NSError)
+      return
+    }
+
+    // TODO: Better typing
+    downloadQueue = HTTPDownloader(self)
+    let commonUserData: [String: Any] = [
+      Key.keyboardInfo: kbInfo,
+      Key.update: isUpdate ? 1 : 0
+    ]
+    downloadQueue!.userInfo = commonUserData
+
+    let kbVersion = keyboardDict[Key.keyboardVersion]
+    let keyboardPath = self.keyboardPath(forFilename: keyboardURL.lastPathComponent,
+                                         keyboardVersion: kbVersion)
+
+    var request = HTTPDownloadRequest(url: keyboardURL, userInfo: commonUserData)
+    request.destinationFile = keyboardPath?.path
+    request.tag = 0
+    downloadQueue!.addRequest(request)
+
+    for (i, url) in kbFontURLs.enumerated() {
+      request = HTTPDownloadRequest(url: url, userInfo: commonUserData)
+      request.destinationFile = fontPath(forFilename: url.lastPathComponent)?.path
+      request.tag = i + 1
+      downloadQueue!.addRequest(request)
+    }
+    downloadQueue!.run()
+  }
+
+  private func keyboardFontURLs(jsonString: String) throws -> [URL] {
+    let font = try? JSONSerialization.jsonObject(with: jsonString.data(using: String.Encoding.utf8)!,
+                                                 options: .mutableContainers)
+
+    guard let fontObject = font as? [AnyHashable: Any] else {
+      return []
+    }
+
+    guard let baseString = options[Key.fontBaseURI] as? String,
+          let baseURL = URL(string: baseString) else {
+      kmLog("Missing font base URL in options: \(options)", checkDebugPrinting: false)
+      return []
+    }
+
+    let fontFiles = fontObject[Key.fontFiles]
+    if let fontFiles = fontFiles as? [String] {
+      return fontFiles.filter({ $0.hasFontExtension }).map({ baseURL.appendingPathComponent($0) })
+    } else if let source = fontFiles as? String {
+      if source.hasFontExtension {
+        return [baseURL.appendingPathComponent(source)]
+      }
+      return []
+    } else {
+      let message = "Unexpected error: \(String(describing: fontFiles)) is not a valid type to be processed."
+      throw NSError(domain: "Keyman", code: 0, userInfo: [NSLocalizedDescriptionKey: message])
+    }
+  }
+
+  private func urlOrThrow(string: String) throws -> URL {
```

Comment:
What's this?  It doesn't seem to be used by Manager.swift, at the least, and it wasn't in the original Obj-C code - or at least, KMManager.m.

Reply:
Helper function that I added but was later unneeded. Will remove.

Reply:
- [ ] Remove unneeded but PR-introduced helper function.